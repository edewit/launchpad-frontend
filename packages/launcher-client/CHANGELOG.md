# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

## [1.3.2](https://github.com/fabric8-launcher/launcher-client/compare/v1.3.1...v1.3.2) (2019-07-08)

**Note:** Version bump only for package @launcher/client
