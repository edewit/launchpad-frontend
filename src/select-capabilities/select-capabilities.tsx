import * as React from "react";
import {useEffect, useState} from "react";
import {
  Button,
  DataList,
  DataListCell,
  DataListCheck,
  DataListContent,
  DataListItem,
  Title,
  Toolbar,
  ToolbarGroup
} from "@patternfly/react-core";
import {FieldEnum} from "./fields/field-enum";

interface Field {
  id: string;
  name: string;
  description: string;
  required: boolean;
  values?: Array<{ id: string, name: string }>
  type: string;
  default?: string
}

interface CapabilityItem {
  id: string;
  name: string;
  description: string;
  category: string;
  fields?: Field[];
  icon?: string;
  selected: boolean;
  data?: any;
  disabled?: boolean;
}

type CapabilityCardProps = CapabilityItem & {
  onChange(changes: { id: string, selected: boolean, data?: any }): void;
}


function CapabilityCard(props: CapabilityCardProps) {
  const onChangeSelected = (selected) => {
    props.onChange({id: props.id, selected, data: props.data})
  };

  const onChangeData = (data) => {
    props.onChange({id: props.id, data, selected: props.selected})
  };
  const elId = `toggle-capability-props-form-${props.id}`;
  const fields = (props.fields || []).filter(f => f.type === 'enum');
  return (
    <DataListItem aria-labelledby={props.id} isExpanded={props.selected}>
      <DataListCheck aria-labelledby={elId} name="Selection item check" onChange={onChangeSelected}
                     checked={props.selected} isDisabled={props.disabled}/>
      <DataListCell width={1} style={{flex: 'none'}}><img src={props.icon}/></DataListCell>
      <DataListCell width={1}><Title size="lg">{props.name}</Title></DataListCell>
      <DataListCell width={3}>{props.description}</DataListCell>
      {fields.length > 0 && props.selected && (
        <DataListContent isHidden={!props.selected} aria-label={`capability-props-form-${props.id}`}>
          {fields.map(f => {
            const selectedValue = (props.data && props.data[f.id]) || f.default;
            const onFieldChange = (v) => {
              const newData = {...props.data, [f.id]: v};
              onChangeData(newData);
            };
            return (
              <FieldEnum key={f.id} parent={props.id} id={f.id} name={f.name} description={f.description}
                         values={f.values!}
                         required={f.required} value={selectedValue}
                         onChange={onFieldChange}
              />
            );
          })}
        </DataListContent>
      )}
    </DataListItem>
  );
}

export interface FieldValue {
  id: string;
  name: string;
  description?: string;
  icon?: string;
  metadata?: object;
}

interface SelectCapabilitiesProps {
  items: CapabilityItem[];

  onSave?(items: CapabilityItem[]);

  onCancel?();
}

export function SelectCapabilities(props: SelectCapabilitiesProps) {

  const [items, setItems] = useState<CapabilityItem[]>(props.items);

  useEffect(() => {
    setItems(props.items);
  }, [props.items]);

  const itemsMap = new Map(items.map(i => [i.id, i] as [string, CapabilityItem]));

  const onChange = (changes: { id: string, selected: boolean, data?: any }) => {
    itemsMap.set(changes.id, {...itemsMap.get(changes.id)!, ...changes});
    setItems(Array.from(itemsMap.values()));
  };

  const onSave = () => {
    if (props.onSave) {
      props.onSave(items);
    }
  };

  const onCancel = () => {
    setItems(props.items);
    if (props.onCancel) {
      props.onCancel();
    }
  };

  return (
    <div className="select-capabilities" style={{padding: '20px'}}>
      <DataList aria-label="select-capability">
        {
          items.map((cap, i) => (
            <CapabilityCard
              {...cap}
              key={i}
              onChange={onChange}
            />
          ))
        }
      </DataList>
      <Toolbar style={{marginTop: '20px'}}>
        <ToolbarGroup>
          <Button variant="primary" onClick={onSave}>Save</Button>
        </ToolbarGroup>
        <ToolbarGroup>
          <Button variant="secondary" onClick={onCancel}>Cancel</Button>
        </ToolbarGroup>
      </Toolbar>
    </div>
  );
}