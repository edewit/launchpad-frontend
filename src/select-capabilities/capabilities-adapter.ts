import {Capability} from "launcher-client";

export function capabilityToItem(c: Capability) {
  return {
    id: c.module,
    name: c.name,
    description: c.description,
    category: c.metadata.category,
    icon: c.metadata.icon,
    fields: c.props,
    selected: c.module === 'welcome',
    disabled: c.module === 'welcome'
  };
}