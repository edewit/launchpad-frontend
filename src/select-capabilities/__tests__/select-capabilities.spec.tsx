import * as React from 'react';
import {cleanup, render} from "react-testing-library";
import {SelectCapabilities} from "../select-capabilities";
import {mockLauncherClient} from 'launcher-client';

const client = mockLauncherClient({creatorUrl: 'efe', launcherURL: 'eqg'});

afterEach(cleanup);

describe('<SelectCapabilities />', () => {
  it('renders the SelectCapabilities correctly', async () => {
    const capabilities = await client.capabilities();
    const items = capabilities.map(c => ({
      id: c.module,
      name: c.name,
      description: c.description,
      category: c.metadata.category,
      icon: c.metadata.icon,
      fields: c.props,
      selected: c.module === 'welcome',
      disabled: c.module === 'welcome'
    })).filter(c => c.category !== 'frontend');
    const comp = render(<SelectCapabilities items={items}/>);
    expect(comp.asFragment()).toMatchSnapshot();
  });
});
