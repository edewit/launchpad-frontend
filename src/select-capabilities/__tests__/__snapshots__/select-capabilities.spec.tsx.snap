// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SelectCapabilities /> renders the HubNSpoke correctly 1`] = `
<DocumentFragment>
  <div
    class="select-capabilities"
  >
    <ul
      aria-label="select-capability"
      class="pf-c-data-list pf-u-box-shadow-md"
      role="list"
    >
      <li
        aria-labelledby="Selection item"
        class="pf-c-data-list__item"
      >
        <div
          class="pf-c-data-list__check"
        >
          <input
            aria-invalid="false"
            aria-labelledby="Selection item check"
            class="pf-c-check__input"
            name="Selection item check"
            type="checkbox"
          />
        </div>
        <div
          class="pf-c-data-list__cell"
        >
          <img />
        </div>
        <div
          class="pf-c-data-list__cell pf-m-flex-2"
        >
          Relational Persistence
        </div>
        <div
          class="pf-c-data-list__cell pf-m-flex-3"
        >
          Sets up basic CRUD (Create, Read, Update, Delete) operations with a relational database
        </div>
      </li>
      <li
        aria-labelledby="Selection item"
        class="pf-c-data-list__item"
      >
        <div
          class="pf-c-data-list__check"
        >
          <input
            aria-invalid="false"
            aria-labelledby="Selection item check"
            class="pf-c-check__input"
            name="Selection item check"
            type="checkbox"
          />
        </div>
        <div
          class="pf-c-data-list__cell"
        >
          <img />
        </div>
        <div
          class="pf-c-data-list__cell pf-m-flex-2"
        >
          HTTP API
        </div>
        <div
          class="pf-c-data-list__cell pf-m-flex-3"
        >
          Exposes an HTTP API for receiving invocations across network boundaries
        </div>
      </li>
      <li
        aria-labelledby="Selection item"
        class="pf-c-data-list__item"
      >
        <div
          class="pf-c-data-list__check"
        >
          <input
            aria-invalid="false"
            aria-labelledby="Selection item check"
            class="pf-c-check__input"
            name="Selection item check"
            type="checkbox"
          />
        </div>
        <div
          class="pf-c-data-list__cell"
        >
          <img />
        </div>
        <div
          class="pf-c-data-list__cell pf-m-flex-2"
        >
          Web Application
        </div>
        <div
          class="pf-c-data-list__cell pf-m-flex-3"
        >
          Adds a simple Hello World web application to the user's project
        </div>
      </li>
      <li
        aria-labelledby="Selection item"
        class="pf-c-data-list__item"
      >
        <div
          class="pf-c-data-list__check"
        >
          <input
            aria-invalid="false"
            aria-labelledby="Selection item check"
            class="pf-c-check__input"
            name="Selection item check"
            type="checkbox"
          />
        </div>
        <div
          class="pf-c-data-list__cell"
        >
          <img />
        </div>
        <div
          class="pf-c-data-list__cell pf-m-flex-2"
        >
          Welcome Application
        </div>
        <div
          class="pf-c-data-list__cell pf-m-flex-3"
        >
          Deploys a Welcome Application where all selected capabilities can be tested. Required.
        </div>
      </li>
    </ul>
    <button
      class="pf-c-button pf-m-primary pf-m-block"
      type="button"
    >
      Save
    </button>
  </div>
</DocumentFragment>
`;

exports[`<SelectCapabilities /> renders the SelectCapabilities correctly 1`] = `
<DocumentFragment>
  <div
    class="select-capabilities"
  >
    <ul
      aria-label="select-capability"
      class="pf-c-data-list pf-u-box-shadow-md"
      role="list"
    >
      <li
        aria-labelledby="Selection item"
        class="pf-c-data-list__item"
      >
        <div
          class="pf-c-data-list__check"
        >
          <input
            aria-invalid="false"
            aria-labelledby="Selection item check"
            class="pf-c-check__input"
            name="Selection item check"
            type="checkbox"
          />
        </div>
        <div
          class="pf-c-data-list__cell"
        >
          <img />
        </div>
        <div
          class="pf-c-data-list__cell pf-m-flex-2"
        >
          Relational Persistence
        </div>
        <div
          class="pf-c-data-list__cell pf-m-flex-3"
        >
          Sets up basic CRUD (Create, Read, Update, Delete) operations with a relational database
        </div>
      </li>
      <li
        aria-labelledby="Selection item"
        class="pf-c-data-list__item"
      >
        <div
          class="pf-c-data-list__check"
        >
          <input
            aria-invalid="false"
            aria-labelledby="Selection item check"
            class="pf-c-check__input"
            name="Selection item check"
            type="checkbox"
          />
        </div>
        <div
          class="pf-c-data-list__cell"
        >
          <img />
        </div>
        <div
          class="pf-c-data-list__cell pf-m-flex-2"
        >
          HTTP API
        </div>
        <div
          class="pf-c-data-list__cell pf-m-flex-3"
        >
          Exposes an HTTP API for receiving invocations across network boundaries
        </div>
      </li>
      <li
        aria-labelledby="Selection item"
        class="pf-c-data-list__item"
      >
        <div
          class="pf-c-data-list__check"
        >
          <input
            aria-invalid="false"
            aria-labelledby="Selection item check"
            class="pf-c-check__input"
            name="Selection item check"
            type="checkbox"
          />
        </div>
        <div
          class="pf-c-data-list__cell"
        >
          <img />
        </div>
        <div
          class="pf-c-data-list__cell pf-m-flex-2"
        >
          Web Application
        </div>
        <div
          class="pf-c-data-list__cell pf-m-flex-3"
        >
          Adds a simple Hello World web application to the user's project
        </div>
      </li>
      <li
        aria-labelledby="Selection item"
        class="pf-c-data-list__item"
      >
        <div
          class="pf-c-data-list__check"
        >
          <input
            aria-invalid="false"
            aria-labelledby="Selection item check"
            class="pf-c-check__input"
            name="Selection item check"
            type="checkbox"
          />
        </div>
        <div
          class="pf-c-data-list__cell"
        >
          <img />
        </div>
        <div
          class="pf-c-data-list__cell pf-m-flex-2"
        >
          Welcome Application
        </div>
        <div
          class="pf-c-data-list__cell pf-m-flex-3"
        >
          Deploys a Welcome Application where all selected capabilities can be tested. Required.
        </div>
      </li>
    </ul>
    <button
      class="pf-c-button pf-m-primary pf-m-block"
      type="button"
    >
      Save
    </button>
  </div>
</DocumentFragment>
`;
